### PR TITLE
Rework searching for unimplemented methods. Fix #161, #182

### LINKED ISSUE #161: Implement methods quickfix on records should not create stubs for equals, hashCode and toString
```java
record MyRecord(Object o) implements Comparable<MyRecord> {
}
```

Invoking the quickfix to add the missing members unexpectedly yields
```java
record MyRecord(Object o) implements Comparable<MyRecord> {

	@Override
	public int compareTo(MyRecord o) {
		return 0;
	}

	@Override
	public boolean equals(Object obj) {
		return false;
	}

	@Override
	public int hashCode() {
		return 0;
	}

	@Override
	public String toString() {
		return null;
	}
	
}
```

### PR DESCRIPTION
**Pull request by tsmaeder**
## What it does
Rework the search for unimplemented methods to be closer to JLS 17
8.4.8. etc. Also replaced feature flags with a method filter for clarity
and extensibility of the algorithm.

Fixes #161 and #182 

## How to test
* creating types via wizards and checking "create unimplemented method" 
* Cleaning up "unimplemented methods"
* "Add unimplemented methods" quickfixes for weird cases

A lot of the cases should be covered by tests, though.

### CODE REVIEW

**Review by jjohnstn — CHANGES_REQUESTED** [review]
Hi Thomas, you need to add test cases to show the code fixes the various issues you have cited.

**Review by jjohnstn — APPROVED** [review]

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java [hunk 1]
```diff
@@ -816,53 +768,133 @@ private static String suggestParameterName(ICompilationUnit unit, IVariableBindi
...
+			if (candidate.getName().equals(method.getName()) && method.isSubsignature(candidate)) {
+				return candidate;
 			}
-			curr= curr.getSuperclass();
 		}
+		return null;
+	}
+
+	private static boolean isStatic(IMethodBinding method) {
+		return Modifier.isPrivate(method.getModifiers());
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java:864]: Ups

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java [hunk 2]
```diff
@@ -816,53 +768,133 @@ private static String suggestParameterName(ICompilationUnit unit, IVariableBindi
...
+	}
+
+	private static boolean isPrivate(IMethodBinding method) {
+		return Modifier.isStatic(method.getModifiers());
```
Comment (stephan-herrmann) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java:868]: Ooops.

**Review by stephan-herrmann — COMMENTED** [review]
@tsmaeder s.o. seems to be confused about `static` vs `private` :)

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java (hunk 2 above)
Comment (jjohnstn) [reply org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java:868]: Thanks Stephan for point this out.  I have created: https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/213
to fix this.

**Pull request merged**